compiler/src/iree/compiler/Codegen/Common/GPU/GPUHeuristics.cpp

```diff
@@ -132,122 +136,254 @@ FailureOr<GPUMMASchedule> fitScheduleInSharedMemory(
+    nTileCount = nGCD.getSExtValue();
+    remainingTiles /= nTileCount;
+
+    APInt mGCD = GreatestCommonDivisor(APInt(64, mTotalTileCount),
+                                       APInt(64, remainingWarps));
+    mWarpCount = mGCD.getSExtValue();
+    mTotalTileCount /= mWarpCount;
+    remainingWarps /= mWarpCount;
+
+    mGCD = GreatestCommonDivisor(APInt(64, mTotalTileCount),
+                                 APInt(64, remainingTiles));
+    mTileCount = mGCD.getSExtValue();
+  }
+
+  const uint64_t kTotalTileCount =
+      llvm::divideCeil(problem.kSize, intrinsic.kSize);
+  APInt kGCD = GreatestCommonDivisor(
+      APInt(64, kTotalTileCount), APInt(64, seeds.bestKTileCountPerSubgroup));
+  int64_t kTileCount = kGCD.getSExtValue();
+
+  return GPUMMASchedule{intrinsicIndex,  intrinsic.mSize, intrinsic.nSize,
+                        intrinsic.kSize, mWarpCount,      nWarpCount,
+                        mTileCount,      nTileCount,      kTileCount};
+}
+
 FailureOr<GPUMMASchedule> deduceMMASchedule(
     const GPUMatmulShapeType &problem, ArrayRef<GPUMatmulShapeType> intrinsics,
     const GPUMMAHeuristicSeeds &seeds, int64_t sharedMemLimitInBytes,
     int64_t subgroupSize, bool transposedLhs, bool transposedRhs,
     bool canUpcastAcc, bool mustBeAligned) {
   for (auto [index, intrinsic] : llvm::enumerate(intrinsics)) {
-    if (problem.aType != intrinsic.aType || problem.bType != intrinsic.bType) {
-      continue; // Cannot use this intrinsic for mismatched types
-    }
-    if (problem.cType != intrinsic.cType) {
-      auto isFpCase =
-          isa<FloatType>(problem.cType) && isa<FloatType>(intrinsic.cType);
-      auto isUpcast = problem.cType.getIntOrFloatBitWidth() <
-                      intrinsic.cType.getIntOrFloatBitWidth();
-      if (!(canUpcastAcc && isFpCase && isUpcast)) {
-        continue; // Cannot use this intrinsic if not upcasting
-      }
+    if (failed(canTargetIntrinsic(problem, intrinsic, canUpcastAcc,
+                                  mustBeAligned))) {
+      continue;
     }
 
-    if (mustBeAligned && (problem.mSize % intrinsic.mSize != 0 ||
-                          problem.nSize % intrinsic.nSize != 0 ||
-                          problem.kSize % intrinsic.kSize != 0)) {
-      continue; // Cannot use this intrinsic for misaligned cases
-    }
+    GPUMMASchedule schedule =
+        getOptimalMMASchedule(problem, intrinsic, seeds, index);
+
+    LLVM_DEBUG({
+      llvm::dbgs() << "chosen MMA schedule:\n";
+      llvm::dbgs() << "  intrinsic (M, N, K) = (" << intrinsic.mSize << ", "
+                   << intrinsic.nSize << ", " << intrinsic.kSize << ")\n";
+      llvm::dbgs() << "  subgroup count (M, N) = (" << schedule.mWarpCount
+                   << ", " << schedule.nWarpCount << ")\n";
+      llvm::dbgs() << "  subgroup tile count (M, N, K) = ("
+                   << schedule.mTileCount << ", " << schedule.nTileCount << ", "
+                   << schedule.kTileCount << ")\n";
```

Comment:
Also here, could you add a printer for intrinsic and use the new one for schedule?

Reply:
I removed the intrinsic printing and just used the schedule printer. the schedule prints the intrinsic size also.